### ISSUE TITLE
E: ioctlsocket FIONBIO 1 failed with error 0

### ISSUE THREAD

**Issue by crazbot** [warmcat/libwebsockets#1747]
this error occurred when i upgraded the libwebsockets to master version, and i'm working on Windows. everything goes well if i turn on the logs, otherwise error occurred

**Comment by lws-team** [comment]
I think this is just bad luck to update to master in the day between getting a patch that accidentally caused this and getting a patch to fix it https://github.com/warmcat/libwebsockets/pull/1739

If you update to today's master it should be gone.

**Comment by crazbot** [comment]
Thank you so much, i'll try it later.

**Comment by crazbot** [comment]
I updated to master today, and the problem still exists. The following piece of the log show WSI_TOKEN_PROTOCOL is null, but the connection succeeded still:

[2019/11/04 10:04:25:8119] D: lws_parse: unk header 23 'sec-websocket-location:'
[2019/11/04 10:04:25:8159] I: lws_client_ws_upgrade: WSI_TOKEN_PROTOCOL is null
[2019/11/04 10:04:25:8179] D: Selected protocol hydra

And if I comment out `lws_set_log_level(LLL_DEBUG | LLL_INFO | LLL_NOTICE | LLL_WARN | LLL_ERR | LLL_CLIENT, NULL);`, the connection hung around

**Comment by lws-team** [comment]
This issue is rambling around a lot.  The title is about a spurious error message, that's gone with the update, right?

Not providing a protocol header is legal, lws will bind it to your first protocol by default.

What does the logging actually care about?  Eg if you call the logging api with LLL_ERR alone it's able to work?  What does 'connection hung around' mean?

**Comment by crazbot** [comment]
It's able to work if I call the logging api with LLL_DEBUG|LLL_INFO|LLL_NOTICE|LLL_WARN|LLL_ERR, otherwise it doesn't work

**Comment by lws-team** [comment]
So "doesn't work" case means "connection hung around"... what does that look like?  It's a websocket connection?  How did you notice it's 'hanging around'?  What steps should I take after I change the logging, to see it?

**Comment by crazbot** [comment]
Yeah, it's a websocket connection. after `lws_client_connect_via_info` called, the callback did not been called with `LWS_CALLBACK_CLIENT_ESTABLISHED` message

**Comment by crazbot** [comment]
I think it's weird that when i launch a debug session and break on `lws_client_connect_via_info`, then step-over, and the connection works. If i break on `lws_client_connect_via_info`, then remove the breakpoint, and continue the execution, the connection doesn't work.

**Comment by crazbot** [comment]
It also works if I turn on the logs.

**Comment by crazbot** [comment]
And it works normally as well, if I connect to the local server. The problem happens only when connect to a remote server

**Comment by lws-team** [comment]
So earlier I was asking what if you just

>  call the logging api with LLL_ERR alone 

You replied it "only" works if you call it with 4 log types.

What happens if you build in release mode (`-DCMAKE_BUILD_TYPE=RELEASE` at cmake)?  This disables most of the logging for build but leaves ERR, WARN, NOTICE and USER.

**Comment by crazbot** [comment]
No log info printed on the console, if called the logging api with LLL_ERR

**Comment by lws-team** [comment]
Yeah but does it fail to get ESTABLISHED?  I am trying to figure out what makes the different behaviour, it sounds related to time.

**Comment by crazbot** [comment]
Yeah it failed everytime if called the logging api with LLL_ERR.

**Comment by lws-team** [comment]
What do the logs say in the fail case if you, eg, enable ERR|WARN|NOTICE|INFO?

**Comment by crazbot** [comment]
Nothing.

**Comment by lws-team** [comment]
Literally no logs are coming out?

**Comment by crazbot** [comment]
Yes.

**Comment by lws-team** [comment]
So all your logs otherwise say I or D or so, none say N?

**Comment by crazbot** [comment]
`lws_set_log_level(LLL_ERR, NULL);` --- this is how i call the log api, and non say I, D, or N

**Comment by lws-team** [comment]
This has gotten pretty confused.  Just using LLL_ERR, you will not see any logs unless some big error is happening in lws.  That is expected.

I suggested to try that to find out if this

> connection hung around

problem is still coming, because you also said

> It also works if I turn on the logs.

So I wanted to know if it works if you "turn on the logs" with just LLL_ERR?  You say

> it failed every time if calling the logs with LLL_ERR

OK... so I suggested next

> What do the logs say in the fail case if you, eg, enable ERR|WARN|NOTICE|INFO?

You said

> Nothing

But from the sound of it, you are still using just LLL_ERR, we expect nothing to come because there are no critical errors.  It sounds like if we slow it down by printing logs is the difference, which log level does it need... is LLL_ERR | LLL_WARN | LLL_NOTICE | LLL_INFO enough?  Or does it need LLL_DEBUG?
